File: website/src/lib/jd-matcher.ts
/**
 * jd-matcher.ts — Composite JD requirement scorer (TypeScript port).
 *
 * Scoring formula:
 *   composite = exact * 0.4 + semantic * 0.3 + metadata * 0.3
 *
 * Thresholds:
 *   >= 0.7  → "met"
 *   >= 0.4  → "partial"
 *   <  0.4  → "gap"
 */

export interface RequirementScore {
  requirement: string;
  requirement_type: string;
  exact_score: number;
  semantic_score: number;
  metadata_score: number;
  composite_score: number;
  status: "met" | "partial" | "gap";
}

// ---------------------------------------------------------------------------
// Cosine similarity
// ---------------------------------------------------------------------------

/**
 * Cosine similarity between two equal-length vectors.
 * Returns 0 if either vector has zero magnitude.
 */
export function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0, normA = 0, normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  if (normA === 0 || normB === 0) return 0;
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

/**
 * Given a requirement embedding and an array of nugget embeddings,
 * return the maximum cosine similarity (0-1).
 */
export function maxSemanticScore(
  reqEmbedding: number[],
  nuggetEmbeddings: number[][]
): number {
  if (nuggetEmbeddings.length === 0) return 0;
  let best = 0;
  for (const ne of nuggetEmbeddings) {
    const sim = cosineSimilarity(reqEmbedding, ne);
    if (sim > best) best = sim;
  }
  return Math.max(0, Math.min(1, best));
}

// ---------------------------------------------------------------------------
// Stop words
// ---------------------------------------------------------------------------

const STOPWORDS = new Set([
  "and", "or", "the", "a", "an", "with", "in", "for", "of", "to",
  "at", "by", "from", "on", "as", "is", "are", "was", "were",
  "be", "been", "have", "has", "had", "will", "would", "could",
  "should", "may", "might", "must", "shall",
]);

const FINTECH_COMPANIES = [
  "amex", "american express", "visa", "mastercard", "stripe", "paypal",
  "bank", "capital", "financial", "goldman", "morgan", "citi", "jpmorgan",
];

const FINTECH_KEYWORDS = ["fintech", "banking", "finance", "crypto", "payments"];

const LEADERSHIP_KEYWORDS = ["lead", "manage", "director", "head of", "vp"];

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function tokenize(text: string): string[] {
  return text
    .replace(/[^a-zA-Z0-9\s]/g, " ")
    .toLowerCase()
    .split(/\s+/)
    .filter((w) => w.length >= 2 && !STOPWORDS.has(w));
}

/**
 * Fuzzy token match: token matches if it appears as a substring in the answer
 * token, or the answer token appears as a substring in the requirement token.
 * This handles: "SQL" → "MySQL"/"PostgreSQL", "Python" → "Python3"/"python-based".
 */
function tokenMatchesFuzzy(reqToken: string, answerToken: string): boolean {
  return answerToken.includes(reqToken) || reqToken.includes(answerToken);
}

// ---------------------------------------------------------------------------
// Exact match score
// ---------------------------------------------------------------------------

/**
 * Score how well a requirement's keywords appear in nugget answers.
 *
 * Improvements over simple substring check:
 * - Full stop word removal (30+ words) so noise tokens don't dilute score
 * - Per-token fuzzy matching: "SQL" matches "MySQL"/"PostgreSQL"/"NoSQL",
 *   "Python" matches "Python3"/"python-based"
 * - Score = matched_tokens / total_tokens, capped at 1.0
 * - Returns best score across all nugget answers
 */
export function exactMatchScore(
  requirement: string,
  nuggetAnswers: string[]
): number {
  const reqTokens = tokenize(requirement);
  if (reqTokens.length === 0) return 0;

  let bestRatio = 0;

  for (const answer of nuggetAnswers) {
    if (!answer) continue;
    // Tokenize answer without stop word filter so all answer words are candidates
    const answerTokens = answer
      .replace(/[^a-zA-Z0-9\s]/g, " ")
      .toLowerCase()
      .split(/\s+/)
      .filter((w) => w.length >= 2);

    let hits = 0;
    for (const reqToken of reqTokens) {
      const matched = answerTokens.some((at) => tokenMatchesFuzzy(reqToken, at));
      if (matched) hits++;
    }

    const ratio = Math.min(hits / reqTokens.length, 1.0);
    if (ratio > bestRatio) bestRatio = ratio;
  }

  return bestRatio;
}

// ---------------------------------------------------------------------------
// Metadata match score
// ---------------------------------------------------------------------------

export interface NuggetMeta {
  section_type: string;
  company?: string | null;
  role?: string | null;
  event_date?: string | null;
  answer?: string | null;
}

/**
 * Score a requirement using structured nugget metadata (not answer text).
 * Returns max across three signals: leadership, experience years, industry.
 *
 * @param requirement     raw requirement text
 * @param requirementType requirement category (unused currently, reserved for future)
 * @param nuggets         raw nugget objects with typed metadata fields
 */
export function metadataMatchScore(
  requirement: string,
  requirementType: string,
  nuggets: NuggetMeta[]
): number {
  const reqLower = requirement.toLowerCase();
  const signals: number[] = [];

  // ── Signal 1: Leadership ───────────────────────────────────────────────────
  // If requirement mentions leadership keywords, check nugget role fields
  const isLeadershipReq = LEADERSHIP_KEYWORDS.some((kw) => reqLower.includes(kw));
  if (isLeadershipReq) {
    const hasLeadershipRole = nuggets.some((n) => {
      const role = (n.role ?? "").toLowerCase();
      return LEADERSHIP_KEYWORDS.some((kw) => role.includes(kw));
    });
    signals.push(hasLeadershipRole ? 1.0 : 0.0);
  }

  // ── Signal 2: Experience years ─────────────────────────────────────────────
  // Extract "N+ years" from requirement, sum work_experience date spans
  const yearsMatch = requirement.match(/(\d+)\+?\s*years?/i);
  if (yearsMatch) {
    const requiredYears = parseInt(yearsMatch[1], 10);
    const currentYear = new Date().getFullYear();
    let totalYears = 0;

    for (const nugget of nuggets) {
      if (nugget.section_type !== "work_experience") continue;
      const eventDate = nugget.event_date;
      if (!eventDate) continue;

      try {
        // event_date format: "YYYY-MM-DD" for start; some nuggets may use
        // "YYYY-MM-DD/YYYY-MM-DD" range or just the start date
        const parts = eventDate.split("/");
        const startYear = parseInt(parts[0].substring(0, 4), 10);
        let endYear: number;
        if (parts.length > 1) {
          const endPart = parts[1].toLowerCase().trim();
          endYear = endPart === "present" ? currentYear : parseInt(endPart.substring(0, 4), 10);
        } else {
          // single date — treat as ongoing if in the past, else skip
          endYear = currentYear;
        }
        if (startYear > 0 && endYear >= startYear) {
          totalYears += endYear - startYear;
        }
      } catch {
        /* skip malformed dates */
      }
    }

    if (totalYears >= requiredYears) signals.push(1.0);
    else if (totalYears >= requiredYears - 1) signals.push(0.7);
    else signals.push(0.0);
  }

  // ── Signal 3: Industry (fintech/finance) ───────────────────────────────────
  // If requirement mentions fintech/finance terms, check company names in nuggets
  const isFinanceReq = FINTECH_KEYWORDS.some((kw) => reqLower.includes(kw));
  if (isFinanceReq) {
    const hasFinanceCompany = nuggets.some((n) => {
      const company = (n.company ?? "").toLowerCase();
      return FINTECH_COMPANIES.some((term) => company.includes(term));
    });
    signals.push(hasFinanceCompany ? 1.0 : 0.0);
  }

  // Return max signal score; 0.0 if no signal matched
  return signals.length > 0 ? Math.max(...signals) : 0.0;
}

// ---------------------------------------------------------------------------
// Weighted composite
// ---------------------------------------------------------------------------

export function weightedComposite(
  exact: number,
  semantic: number,
  metadata: number
): number {
  const EXACT_WEIGHT = 0.4;
  const SEMANTIC_WEIGHT = 0.3;
  const METADATA_WEIGHT = 0.3;
  return exact * EXACT_WEIGHT + semantic * SEMANTIC_WEIGHT + metadata * METADATA_WEIGHT;
}

// ---------------------------------------------------------------------------
// Batch scorer
// ---------------------------------------------------------------------------

/**
 * Score a list of requirements against nugget answer strings.
 * For full metadata scoring, use scoreRequirementsWithNuggets instead.
 *
 * @param requirements  array of {text, type}
 * @param nuggetAnswers flat array of answer strings from all nuggets
 * @param semanticScores optional map of requirement text → semantic score (0-1)
 */
export function scoreRequirements(
  requirements: { text: string; type: string }[],
  nuggetAnswers: string[],
  semanticScores?: Record<string, number>
): RequirementScore[] {
  return requirements.map((req) => {
    const exact = exactMatchScore(req.text, nuggetAnswers);
    const semantic = semanticScores?.[req.text] ?? 0;
    const metadata = 0; // no raw nuggets available in this variant
    const composite = weightedComposite(exact, semantic, metadata);
    const status: "met" | "partial" | "gap" =
      composite >= 0.7 ? "met" : composite >= 0.4 ? "partial" : "gap";

    return {
      requirement: req.text,
      requirement_type: req.type,
      exact_score: exact,
      semantic_score: semantic,
      metadata_score: metadata,
      composite_score: composite,
      status,
    };
  });
}

/**
 * Full scorer that includes metadata signals.
 *
 * @param requirements  array of {text, type}
 * @param nuggets       typed nugget objects with section_type, company, role, event_date, answer
 * @param semanticScores optional map of requirement text → semantic score (0-1)
 */
export function scoreRequirementsWithNuggets(
  requirements: { text: string; type: string }[],
  nuggets: NuggetMeta[],
  semanticScores?: Record<string, number>
): RequirementScore[] {
  const nuggetAnswers = nuggets.map((n) => n.answer ?? "").filter(Boolean);

  return requirements.map((req) => {
    const exact = exactMatchScore(req.text, nuggetAnswers);
    const semantic = semanticScores?.[req.text] ?? 0;
    const metadata = metadataMatchScore(req.text, req.type, nuggets);
    const composite = weightedComposite(exact, semantic, metadata);
    const status: "met" | "partial" | "gap" =
      composite >= 0.7 ? "met" : composite >= 0.4 ? "partial" : "gap";

    return {
      requirement: req.text,
      requirement_type: req.type,
      exact_score: exact,
      semantic_score: semantic,
      metadata_score: metadata,
      composite_score: composite,
      status,
    };
  });
}

File: context/cli/linkright/src/linkright/tools/jd_matcher.py
"""
jd_matcher.py — Composite JD requirement scorer for LinkRight.

Scoring formula:
    composite = exact_match * 0.4 + semantic_match * 0.3 + metadata_match * 0.3

Thresholds:
    >= 0.7  → "met"
    >= 0.4  → "partial"
    <  0.4  → "gap"
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional


# ---------------------------------------------------------------------------
# Data model
# ---------------------------------------------------------------------------

@dataclass
class RequirementScore:
    requirement: str
    requirement_type: str  # hard_skill | soft_skill | experience | certification | education
    exact_score: float = 0.0
    semantic_score: float = 0.0
    metadata_score: float = 0.0
    composite_score: float = 0.0
    status: str = "gap"  # "met" | "partial" | "gap"
    matching_nuggets: list[str] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Stop words
# ---------------------------------------------------------------------------

_STOPWORDS: frozenset[str] = frozenset(
    ["and", "or", "the", "a", "with", "in", "for", "of", "to"]
)

# ---------------------------------------------------------------------------
# Exact match
# ---------------------------------------------------------------------------

def _tokenize(text: str) -> list[str]:
    """Lower-case, strip punctuation, remove stop words, return non-empty tokens."""
    words = re.sub(r"[^a-zA-Z0-9\s+]", " ", text).lower().split()
    return [w for w in words if w not in _STOPWORDS and len(w) >= 2]


def _keyword_in_text(keyword: str, text: str) -> bool:
    """
    Return True if *keyword* appears in *text* with fuzzy tolerance:
    - exact substring match (case-insensitive), OR
    - keyword is contained within a longer word (e.g. 'sql' in 'mysql').
    """
    text_lower = text.lower()
    kw_lower = keyword.lower()
    return kw_lower in text_lower


def exact_match_score(requirement: str, nuggets: list[dict]) -> float:
    """
    Score how well a requirement's keywords appear verbatim (or as substrings)
    across the user's nuggets.

    Returns the best ratio found in any single nugget: 0.0 – 1.0.
    """
    keywords = _tokenize(requirement)
    if not keywords:
        return 0.0

    best_ratio = 0.0
    for nugget in nuggets:
        answer: str = nugget.get("answer", "") or ""
        if not answer:
            continue
        hits = sum(1 for kw in keywords if _keyword_in_text(kw, answer))
        ratio = hits / len(keywords)
        if ratio > best_ratio:
            best_ratio = ratio

    return best_ratio


# ---------------------------------------------------------------------------
# Metadata match
# ---------------------------------------------------------------------------

_FINTECH_TERMS: frozenset[str] = frozenset(
    ["bank", "finance", "capital", "amex", "visa", "mastercard",
     "stripe", "paypal", "fintech", "fintec"]
)

_LEADERSHIP_SIGNALS: frozenset[str] = frozenset(
    ["team_lead", "manager", "director"]
)

_YEARS_RE = re.compile(r"(\d+)\+?\s*(?:years?|yrs?)", re.IGNORECASE)


def _extract_required_years(requirement: str) -> Optional[int]:
    """Extract the minimum years mentioned in a requirement string."""
    m = _YEARS_RE.search(requirement)
    return int(m.group(1)) if m else None


def metadata_match_score(
    requirement: str, requirement_type: str, nuggets: list[dict]
) -> float:
    """
    Score a requirement using structured metadata fields in nuggets.

    Supported signals:
    - leadership_signal field for leadership requirements
    - event_date spans for experience-year requirements
    - company name keywords for industry (fintech/finance) requirements

    Returns 0.0 – 1.0.
    """
    req_lower = requirement.lower()

    # ── Leadership ───────────────────────────────────────────────────────────
    if any(word in req_lower for word in ("lead", "leadership", "manager", "director")):
        for nugget in nuggets:
            signal = nugget.get("leadership_signal", "")
            if signal in _LEADERSHIP_SIGNALS:
                return 1.0
        return 0.0

    # ── Experience years ─────────────────────────────────────────────────────
    required_years = _extract_required_years(requirement)
    if required_years is not None:
        # Sum durations from work_experience nuggets that carry event_date spans.
        # event_date is expected as a dict {"start": YYYY, "end": YYYY | "present"}
        import datetime
        current_year = datetime.date.today().year
        total_years = 0.0

        for nugget in nuggets:
            if nugget.get("nugget_type") != "work_experience":
                continue
            event_date = nugget.get("event_date") or {}
            if not isinstance(event_date, dict):
                continue
            try:
                start = int(event_date.get("start", 0))
                end_raw = event_date.get("end", "present")
                end = current_year if str(end_raw).lower() == "present" else int(end_raw)
                if start > 0 and end >= start:
                    total_years += end - start
            except (TypeError, ValueError):
                continue

        if total_years >= required_years:
            return 1.0
        if total_years >= required_years - 1:
            return 0.7
        return 0.0

    # ── Fintech / Finance industry ───────────────────────────────────────────
    if any(term in req_lower for term in ("fintech", "banking", "finance", "financial")):
        for nugget in nuggets:
            company: str = (nugget.get("company") or nugget.get("organization") or "").lower()
            if any(ft in company for ft in _FINTECH_TERMS):
                return 1.0
        return 0.0

    return 0.0


# ---------------------------------------------------------------------------
# Composite weighting
# ---------------------------------------------------------------------------

def weighted_composite(exact: float, semantic: float, metadata: float) -> float:
    """Combine exact + semantic + metadata scores with configurable weights."""
    EXACT_WEIGHT = 0.4
    SEMANTIC_WEIGHT = 0.3
    METADATA_WEIGHT = 0.3
    return exact * EXACT_WEIGHT + semantic * SEMANTIC_WEIGHT + metadata * METADATA_WEIGHT


# ---------------------------------------------------------------------------
# Single-requirement scorer
# ---------------------------------------------------------------------------

def score_requirement(
    requirement: str,
    requirement_type: str,
    nuggets: list[dict],
    embeddings_map: Optional[dict] = None,
) -> RequirementScore:
    """Score a single JD requirement against the user's nuggets."""
    score = RequirementScore(requirement=requirement, requirement_type=requirement_type)
    score.exact_score = exact_match_score(requirement, nuggets)
    score.metadata_score = metadata_match_score(requirement, requirement_type, nuggets)
    # semantic_score: filled in by caller when embeddings are available
    if embeddings_map:
        score.semantic_score = embeddings_map.get(requirement, 0.0)
    score.composite_score = weighted_composite(
        score.exact_score, score.semantic_score, score.metadata_score
    )
    if score.composite_score >= 0.7:
        score.status = "met"
    elif score.composite_score >= 0.4:
        score.status = "partial"
    else:
        score.status = "gap"
    return score


# ---------------------------------------------------------------------------
# Batch scorer
# ---------------------------------------------------------------------------

def batch_score_requirements(
    requirements: list[dict],
    nuggets: list[dict],
    embeddings_map: Optional[dict] = None,
) -> list[RequirementScore]:
    """
    Score a list of requirements against nuggets.

    Args:
        requirements: list of {"text": str, "type": str}
        nuggets:      user's career nuggets as dicts
        embeddings_map: optional {requirement_text: semantic_score float}

    Returns:
        list of RequirementScore (one per requirement)
    """
    results: list[RequirementScore] = []
    for req in requirements:
        text: str = req.get("text", "")
        req_type: str = req.get("type", "other")
        results.append(
            score_requirement(
                requirement=text,
                requirement_type=req_type,
                nuggets=nuggets,
                embeddings_map=embeddings_map,
            )
        )
    return results
